src/ray/common/scheduling/scheduling_alg.h

```diff
@@ -0,0 +1,147 @@
+#ifndef RAY_COMMON_SCHEDULING_SCHEDULING_H
+#define RAY_COMMON_SCHEDULING_SCHEDULING_H
+
+#include <iostream>
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+
+/// Maximum number of predefined resources.
+#define NUM_PREDIFINED_RESOURCES 4
```

Comment:
The canonical way to do this is to add an element to the end of the enum, i.e., `enum PredefinedResources {CPU = 0, MEM = 1, GPU = 2, TPU = 3, PredefinedResources_MAX}`. That way only one definition needs to be changed when adding a new resource.

Reply:
Hmm, this would be a bit risky. If the user does:

PredefinedResources {CPU = 3, MEM = 2, GPU = 1, TPU = 0, PredefinedResources_MAX}

the value of PredefinedResources_MAX is 1.

Reply:
Ah, yeah you shouldn't be setting the values explicitly at all:
`enum {CPU, MEM, GPU, TPU, PredefinedResources_MAX}`